<a href="https://colab.research.google.com/github/iambrookedrake/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Brooke_Drake_DSPT6_U1S2M1_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [2]:
### YOUR CODE STARTS HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/ --no-check-certificate


--2020-04-09 01:07:06--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 1602 (1.6K) [text/html]
Saving to: ‘index.html’

index.html          100%[===================>]   1.56K  --.-KB/s    in 0s      

2020-04-09 01:07:06 (47.5 MB/s) - ‘index.html’ saved [1602/1602]



In [3]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data --no-check-certificate

--2020-04-09 01:11:32--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-04-09 01:11:32 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind
import matplotlib.pyplot as plt
import seaborn as sns

In [10]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.names --no-check-certificate

--2020-04-09 01:25:20--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.names
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 6868 (6.7K) [application/x-httpd-php]
Saving to: ‘house-votes-84.names’

house-votes-84.name 100%[===================>]   6.71K  --.-KB/s    in 0s      

2020-04-09 01:25:21 (114 MB/s) - ‘house-votes-84.names’ saved [6868/6868]



In [0]:
columns = ['party', 'handicapped-infants', 'water-project-cost-sharing', 
           'adoption-of-the-budget-resolution', 'physician-fee-freeze', 
           'el-salvador-aid', 'religious-groups-in-schools', 
           'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
           'immigration', 'synfuels-corporation-cutback', 'education-spending', 
           'superfund-right-to-sue', 'crime', 'duty-free-exports', 
           'export-administration-act-south-africa']

In [69]:
votes = pd.read_csv('house-votes-84.data', header=None, names=columns)
votes = votes.replace({'y':1, 'n':0})
votes = votes.replace('?', np.nan)
print(votes.shape)
votes.head()

(435, 17)


party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [70]:
rep = votes[votes['party']=='republican']
print(rep.shape)
rep.head()

(168, 17)


party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [71]:
dem = votes[votes['party']=='democrat']
print(dem.shape)
dem.head()

(267, 17)


party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

In [72]:
votes.dtypes

party                                      object
handicapped-infants                       float64
water-project-cost-sharing                float64
adoption-of-the-budget-resolution         float64
physician-fee-freeze                      float64
el-salvador-aid                           float64
religious-groups-in-schools               float64
anti-satellite-test-ban                   float64
aid-to-nicaraguan-contras                 float64
mx-missile                                float64
immigration                               float64
synfuels-corporation-cutback              float64
education-spending                        float64
superfund-right-to-sue                    float64
crime                                     float64
duty-free-exports                         float64
export-administration-act-south-africa    float64
dtype: object

In [74]:
ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

In [0]:
def p_value(dem, rep):
   print(ttest_ind(dem, rep, nan_policy='omit')[1])

In [0]:
def stat(dem, rep):
   print(ttest_ind(dem, rep, nan_policy='omit')[0])

In [79]:
for column in dem:
  if column !='party': 
    p_value(dem[column], rep[column])

1.613440327937243e-18
0.9291556823993485
2.0703402795404463e-77
1.994262314074344e-177
5.600520111729011e-68
2.3936722520597287e-20
8.521033017443867e-31
2.82471841372357e-54
5.03079265310811e-47
0.08330248490425066
1.5759322301054064e-15
1.8834203990450192e-64
1.2278581709672758e-34
9.952342705606092e-47
5.997697174347365e-32
3.652674361672226e-11


In [84]:
for column in dem:
  if column !='party': 
    stat(dem[column], rep[column])

9.205264294809222
-0.08896538137868286
23.21277691701378
-49.36708157301406
-21.13669261173219
-9.737575825219457
12.526187929077842
18.052093200819733
16.437503268542994
-1.7359117329695164
8.293603989407588
-20.500685724563073
-13.51064251060933
-16.342085656197696
12.853146132542978
6.849454815841208


In [105]:
for column in dem:
  if column !='party': 
    stat(dem[column], rep[column]), p_value(dem[column], rep[column])

9.205264294809222
1.613440327937243e-18
-0.08896538137868286
0.9291556823993485
23.21277691701378
2.0703402795404463e-77
-49.36708157301406
1.994262314074344e-177
-21.13669261173219
5.600520111729011e-68
-9.737575825219457
2.3936722520597287e-20
12.526187929077842
8.521033017443867e-31
18.052093200819733
2.82471841372357e-54
16.437503268542994
5.03079265310811e-47
-1.7359117329695164
0.08330248490425066
8.293603989407588
1.5759322301054064e-15
-20.500685724563073
1.8834203990450192e-64
-13.51064251060933
1.2278581709672758e-34
-16.342085656197696
9.952342705606092e-47
12.853146132542978
5.997697174347365e-32
6.849454815841208
3.652674361672226e-11


In [98]:
votes.columns

Index(['party', 'handicapped-infants', 'water-project-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-groups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

#2
Based off the t-stat of 23.2 and p-value of 2.07e-77, it seems water-project-cost-sharing was an issues Democrats favored over Republicans

#3
Based off the t-stat of -49.36 and p-value of 1.99e-177, adoption-of-the-budget-resolution was more favored by Republicans than Democrats.

#4 
From The t-stat of -0.089 and p-value of 0.93, I would conclude that water-project-cost-sharing has bipartisan support


## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!